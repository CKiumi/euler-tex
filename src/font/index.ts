import { INTER_ATOM_SPACING, SIGMAS, AtomKind } from "./sigma";
import { Font, SPEC } from "./spec";
import METRICS from "./data";
export { SPEC, SIGMAS };
export type { Font, AtomKind };

export type CharMetric = {
  depth: number;
  height: number;
  italic: number;
  skew: number;
  width: number;
};

export const getCharMetrics = (char: string, font: Font): CharMetric => {
  try {
    const metrics = METRICS[font][char.charCodeAt(0)];
    return {
      depth: metrics[0],
      height: metrics[1],
      italic: metrics[2],
      skew: metrics[3],
      width: metrics[4],
    };
  } catch (error) {
    throw new Error(`Font metrics not found for font: ${font}.`);
  }
};

export const getSigma = (name: keyof typeof SIGMAS): number => {
  return SIGMAS[name][0];
};

export const getSpacing = (prevKind: AtomKind, curKind: AtomKind): number =>
  (INTER_ATOM_SPACING[prevKind]?.[curKind] ?? 0) * CSSEmPerMu;

const CSSEmPerMu = SIGMAS.quad[0] / 18;
