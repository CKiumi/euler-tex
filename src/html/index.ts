export * from "./svg/inner";
export * from "./svg/pathNode";
export * from "./svg/sqrt";

export const html = <K extends keyof HTMLElementTagNameMap>(
  tagName: K,
  options: {
    text?: string;
    cls?: string[];
    id?: string;
    style?: Partial<CSSStyleDeclaration>;
    children?: HTMLElement[];
  }
): HTMLElementTagNameMap[K] => {
  const span = document.createElement(tagName);
  Object.assign(span.style, options.style);
  options.id && (span.id = options.id);
  options.cls && span.classList.add(...options.cls);
  options.text && (span.innerText = options.text);
  options.children && span.append(...options.children);
  return span;
};
