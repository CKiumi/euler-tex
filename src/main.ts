import "./style.css";
import katex from "katex";
import "katex/dist/katex.min.css";
const app = document.querySelector<HTMLDivElement>("#app");
if (app) {
  app.innerHTML = "<h1>Euler Tex</h1>";
  katex.render("c = \\pm\\sqrt{a^2 + b^2}", app, {
    throwOnError: false,
  });
}
