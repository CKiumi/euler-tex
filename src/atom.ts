import { Box, HBox, SymBox, toVBox, VStackBox } from "/src/box";
import { getCharMetrics, getSigma, getSpacing } from "/font";
import { AtomKind } from "/font/src/sigma";
import { Font } from "/font/src/spec";

export interface Atom {
  type: "sym" | "accent" | "overline" | "line";
  kind: AtomKind;
}

export interface SymAtom extends Atom {
  type: "sym";
  char: string;
  font: Font;
}

export interface AccentAtom extends Atom {
  type: "accent";
  body: Atom;
  accent: Accent;
}

export interface Accent extends SymAtom {
  char: "^" | "~";
  font: "Main-R";
  kind: "ord";
}

export interface OverlineAtom extends Atom {
  type: "overline";
  body: Atom;
}
export interface LineAtom extends Atom {
  type: "line";
}

export const parseAtoms = (atoms: Atom[]): HBox => {
  let prevKind: AtomKind | undefined;
  const children = atoms.map((atom) => {
    const box = parseAtom(atom);
    if (prevKind && atom.kind) {
      box.spacing = getSpacing(prevKind, atom.kind);
    }
    prevKind = atom.kind;
    return box;
  });
  const width = children.reduce((acc, a) => acc + a.width, 0);
  const depth = Math.max(...children.map((child) => child.depth));
  const height = Math.max(...children.map((child) => child.height));
  return { children, width, height, depth };
};

export const parseAtom = (atom: Atom): Box => {
  if (atom.type === "sym") return parseSymAtom(atom as SymAtom);
  if (atom.type === "overline") return parseOverline(atom as OverlineAtom);
  if (atom.type === "line") return parseLine();
  if (atom.type === "accent") return parseAccentAtom(atom as AccentAtom);
  throw new Error("No Atom Type Specified");
};

export const parseSymAtom = ({ char, font }: SymAtom): SymBox => {
  const { depth, height, italic, width } = getCharMetrics(char, font);
  return { char, font, depth, height, width: width + italic, italic };
};

export const parseAccentAtom = (atom: AccentAtom): VStackBox => {
  const box = parseSymAtom(atom.body as SymAtom);
  const accBox = parseSymAtom(atom.accent);
  const clearance = Math.min(box.height, getSigma("xHeight"));
  accBox.spacingBelow = -clearance;
  return toVBox([accBox, box], box.depth);
};

export const parseOverline = (body: OverlineAtom): VStackBox => {
  const accBox = parseAtom({ type: "line", kind: "ord" });
  const box = parseSymAtom(body.body as SymAtom);
  const defaultRuleThickness = getSigma("defaultRuleThickness");
  accBox.spacingTop = defaultRuleThickness;
  accBox.spacingBelow = 3 * defaultRuleThickness;
  return toVBox([accBox, box], box.depth);
};

export const parseLine = (): Box => {
  return { width: 0, height: getSigma("defaultRuleThickness"), depth: 0 };
};
