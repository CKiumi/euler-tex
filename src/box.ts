import { Atom, SymAtom } from "./atom";
import { getCharMetrics } from "/font";
import { Font } from "/font/src/spec";

export interface Box {
  height: number;
  depth: number;
  width: number;
}

export interface SymBox extends Box {
  char: string;
  font: Font;
  height: number;
  depth: number;
  width: number;
  italic: number;
}

export interface HBox extends Box {
  children: Box[];
  height: number;
  depth: number;
  width: number;
}

export const toHBox = (atoms: Atom[]): HBox => {
  const children = atoms.map(toSymBox);
  const width = children.reduce((acc, a) => acc + a.width, 0);
  const depth = Math.max(...children.map((child) => child.depth));
  const height = Math.max(...children.map((child) => child.height));
  return { children, width, height, depth };
};

export const toSymBox = ({ char, font }: SymAtom): SymBox => {
  const { depth, height, italic, width } = getCharMetrics(char, font);
  return { char, font, depth, height, width: width + italic, italic };
};
