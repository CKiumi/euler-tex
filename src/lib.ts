import { GroupAtom } from "./atom/atom";
import { parse } from "./parser/parser";
export * from "./font";
export * from "./atom/atom";
export * from "./box/box";
export * from "./parser/parser";

export const latexToHtml = (latex: string) =>
  new GroupAtom(parse(latex)).toBox().toHtml();

export const latexToEditableAtom = (latex: string) => {
  const atom = new GroupAtom(parse(latex, true), true);
  atom.toBox().toHtml();
  return atom;
};
