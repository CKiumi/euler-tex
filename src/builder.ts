import { HBox, SymBox } from "./box";
import { em } from "./util";
import { SPEC } from "/font/src/spec";

export const buildHBox = (hbox: HBox) => {
  const span = document.createElement("span");
  span.classList.add("hbox");
  hbox.children.forEach((box) => {
    span.append(buildSymBox(box as SymBox));
  });
  return span;
};

export const buildSymBox = ({
  char,
  font,
  height,
  depth,
  italic,
  spacing,
}: SymBox): HTMLSpanElement => {
  const span = document.createElement("span");
  span.innerText = char;
  span.classList.add("box", font.toLowerCase());
  span.style.height = em(height + depth);
  if (italic) {
    span.style.paddingRight = em(italic);
  }
  if (spacing) {
    span.style.marginLeft = em(spacing);
  }
  span.style.lineHeight = em(
    (height + (SPEC[font].descent - SPEC[font].ascent) / 2) * 2
  );
  return span;
};
