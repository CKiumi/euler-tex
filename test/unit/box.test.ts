import { HBox, RectBox, SymBox, VBox, VStackBox } from "/src/lib";
import { expect, test } from "vitest";

const a = new SymBox("a", "Math-I");
const hbox = new HBox([a, a]);
const vStackbox = new VStackBox([a, hbox], 3);
const rectBox = new RectBox({ height: 1, depth: 0.5, width: 2 });

test("symbol box", () => {
  expect(a.toHtml()).toMatchSnapshot();
});

test("hBox box", () => {
  expect(hbox.toHtml()).toMatchSnapshot();
});

test("vBox box", () => {
  const vbox = new VBox([
    { box: a, shift: 1 },
    { box: hbox, shift: -0.5 },
  ]);
  expect(vbox.toHtml()).toMatchSnapshot();
  const vbox2 = new VBox([{ box: a, shift: 1 }]);
  expect(vbox2.toHtml()).toMatchSnapshot();
});

test("vStackBox box", () => {
  expect(vStackbox.toHtml()).toMatchSnapshot();
});

test("rect box", () => {
  expect(rectBox.toHtml()).toMatchSnapshot();
});
