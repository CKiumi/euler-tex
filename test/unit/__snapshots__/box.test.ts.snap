// Vitest Snapshot v1

exports[`accent box 1`] = `
<span
  class="hbox"
>
  <span
    class="vlist"
  >
    <span
      class="box math-i"
      style="height: 0.4306em; line-height: 0.3621em;"
    >
      a
    </span>
    <span
      class="box main-r"
      style="margin-bottom: -0.4306em; height: 0.6944em; line-height: 0.7579em;"
    >
      ^
    </span>
  </span>
  <span
    class="vlist"
  >
    <span
      class="box math-i"
      style="height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
    >
      f
    </span>
    <span
      style="margin-bottom: 0.1200em; margin-top: 0.0400em; height: 0.0400em; background: black; width: 100%;"
    />
  </span>
  <span
    class="vlist"
  >
    <span
      class="box math-i"
      style="height: 0.6833em; padding-right: 0.0715em; line-height: 0.8677em;"
    >
      K
    </span>
    <span
      class="box main-r"
      style="margin-bottom: -0.4310em; height: 0.6679em; line-height: 0.0047em;"
    >
      ~
    </span>
  </span>
  <span
    class="vlist"
  >
    <span
      class="box size2"
      style="height: 2.2222em; padding-right: 0.4445em; line-height: 2.2220em;"
    >
      ∫
    </span>
    <span
      class="box main-r"
      style="margin-bottom: -0.4310em; height: 0.6944em; line-height: 0.7579em;"
    >
      ^
    </span>
  </span>
</span>
`;

exports[`frac 1`] = `
<span
  class="hbox"
>
  <span
    class="vlist"
    style="height: 2.2519em; vertical-align: -0.6860em;"
  >
    <span
      class="hbox"
      style="bottom: 0.0000em;"
    >
      <span
        class="box math-i"
        style="height: 0.6833em; padding-right: 0.0715em; line-height: 0.8677em;"
      >
        K
      </span>
      <span
        class="box main-r"
        style="margin-left: 0.2222em; height: 0.6667em; line-height: 0.5357em;"
      >
        +
      </span>
      <span
        class="box math-i"
        style="margin-left: 0.2222em; height: 0.8540em; padding-right: 0.0572em; line-height: 0.8200em;"
      >
        j
      </span>
    </span>
    <span
      style="height: 0.0400em; background: black; width: 100%; bottom: 0.2327em;"
    />
    <span
      class="hbox"
      style="bottom: 0.4452em;"
    >
      <span
        class="box math-i"
        style="height: 0.4306em; line-height: 0.3621em;"
      >
        a
      </span>
      <span
        class="box main-r"
        style="margin-left: 0.2222em; height: 0.6667em; line-height: 0.5357em;"
      >
        +
      </span>
      <span
        class="box math-i"
        style="margin-left: 0.2222em; height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
      >
        f
      </span>
    </span>
  </span>
</span>
`;

exports[`hbox 1`] = `
<span
  class="hbox"
>
  <span
    class="box math-i"
    style="height: 0.4306em; line-height: 0.3621em;"
  >
    a
  </span>
  <span
    class="box main-r"
    style="margin-left: 0.2222em; height: 0.6667em; line-height: 0.5357em;"
  >
    +
  </span>
  <span
    class="box math-i"
    style="margin-left: 0.2222em; height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
  >
    f
  </span>
  <span
    class="box main-r"
    style="margin-left: 0.2778em; height: 0.2337em; line-height: 0.1027em;"
  >
    =
  </span>
  <span
    class="box size2"
    style="margin-left: 0.2778em; height: 2.2222em; padding-right: 0.4445em; line-height: 2.2220em;"
  >
    ∫
  </span>
</span>
`;

exports[`leftright 1`] = `
<span
  class="hbox"
>
  <span
    class="hbox"
  >
    <span
      class="box main-r"
      style="height: 1.0000em; line-height: 0.8690em;"
    >
      (
    </span>
    <span
      class="hbox"
    >
      <span
        class="box math-i"
        style="height: 0.4306em; line-height: 0.3621em;"
      >
        a
      </span>
      <span
        class="box main-r"
        style="margin-left: 0.2222em; height: 0.6667em; line-height: 0.5357em;"
      >
        +
      </span>
      <span
        class="box math-i"
        style="margin-left: 0.2222em; height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
      >
        f
      </span>
    </span>
    <span
      class="box main-r"
      style="height: 1.0000em; line-height: 0.8690em;"
    >
      )
    </span>
  </span>
  <span
    class="hbox"
    style="margin-left: 0.1667em;"
  >
    <span
      class="box size3"
      style="height: 2.4000em; line-height: 2.4000em;"
    >
      (
    </span>
    <span
      class="hbox"
    >
      <span
        class="box size2"
        style="height: 2.2222em; padding-right: 0.4445em; line-height: 2.2220em;"
      >
        ∫
      </span>
      <span
        class="box main-r"
        style="margin-left: 0.1667em; height: 0.6667em; line-height: 0.5357em;"
      >
        +
      </span>
      <span
        class="box math-i"
        style="height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
      >
        f
      </span>
    </span>
    <span
      class="box size3"
      style="height: 2.4000em; line-height: 2.4000em;"
    >
      )
    </span>
  </span>
</span>
`;

exports[`matrix 1`] = `
<span
  class="hbox"
>
  <span
    class="hbox"
  >
    <span
      class="box size3"
      style="height: 2.4000em; line-height: 2.4000em;"
    >
      (
    </span>
    <span
      class="hbox"
    >
      <span
        class="hbox"
      >
        <span
          class="vlist"
          style="margin-right: 0.5000em; height: 1.6306em; vertical-align: -0.5900em;"
        >
          <span
            class="box math-i"
            style="height: 0.4306em; line-height: 0.3621em;"
          >
            a
          </span>
          <span
            class="box math-i"
            style="height: 0.4306em; line-height: 0.3621em; bottom: 0.7694em;"
          >
            a
          </span>
        </span>
        <span
          class="vlist"
          style="margin-left: 0.5000em; height: 1.6306em; vertical-align: -0.5900em;"
        >
          <span
            class="box math-i"
            style="height: 0.4306em; line-height: 0.3621em;"
          >
            a
          </span>
          <span
            class="box math-i"
            style="height: 0.4306em; line-height: 0.3621em; bottom: 0.7694em;"
          >
            a
          </span>
        </span>
      </span>
    </span>
    <span
      class="box size3"
      style="height: 2.4000em; line-height: 2.4000em;"
    >
      )
    </span>
  </span>
  <span
    class="hbox"
    style="margin-left: 0.1667em;"
  >
    <span
      class="box size3"
      style="height: 2.4000em; line-height: 2.4000em;"
    >
      (
    </span>
    <span
      class="hbox"
    >
      <span
        class="hbox"
      >
        <span
          class="vlist"
          style="margin-right: 0.5000em; height: 1.6306em; vertical-align: -0.5900em;"
        >
          <span
            class="box math-i"
            style="height: 0.4306em; line-height: 0.3621em;"
          >
            a
          </span>
          <span
            class="box math-i"
            style="height: 0.4306em; line-height: 0.3621em; bottom: 0.7694em;"
          >
            a
          </span>
        </span>
        <span
          class="vlist"
          style="margin-left: 0.5000em; height: 0.4306em; vertical-align: -0.5900em;"
        >
          <span
            class="box math-i"
            style="height: 0.4306em; line-height: 0.3621em;"
          >
            a
          </span>
        </span>
      </span>
    </span>
    <span
      class="box size3"
      style="height: 2.4000em; line-height: 2.4000em;"
    >
      )
    </span>
  </span>
</span>
`;

exports[`sqrt 1`] = `
<span
  class="hbox"
>
  <span
    class="vlist"
    style="height: 1.0000em; vertical-align: -0.1908em;"
  >
    <span
      class="hbox"
      style="margin-left: 0.7770em; bottom: 0.1908em;"
    >
      <span
        class="box math-i"
        style="height: 0.4306em; line-height: 0.3621em;"
      >
        a
      </span>
    </span>
    <span
      style="min-width: 0.853em; height: 1.0000em; background: #31121355; display: inline-block; bottom: -0.4306em;"
    >
      <svg
        height="1.0000em"
        preserveAspectRatio="xMinYMin slice"
        viewBox="75 73 400000 1000"
        width="1.30559em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M95,702
  c-2.7,0,-7.17,-2.7,-13.5,-8c-5.8,-5.3,-9.5,-10,-9.5,-14
  c0,-2,0.3,-3.3,1,-4c1.3,-2.7,23.83,-20.7,67.5,-54
  c44.2,-33.3,65.8,-50.3,66.5,-51c1.3,-1.3,3,-2,5,-2c4.7,0,8.7,3.3,12,10
  s173,378,173,378c0.7,0,35.3,-71,104,-213c68.7,-142,137.5,-285,206.5,-429
  c69,-144,104.5,-217.7,106.5,-221
  l0 -0
  c5.3,-9.3,12,-14,20,-14
  H400000v40H845.2724
  s-225.272,467,-225.272,467s-235,486,-235,486c-2.7,4.7,-9,7,-19,7
  c-6,0,-10,-1,-12,-3s-194,-422,-194,-422s-65,47,-65,47z
  M834 80h400000v40h-400000z"
        />
      </svg>
    </span>
  </span>
  <span
    class="vlist"
    style="height: 1.0000em; vertical-align: -0.0228em;"
  >
    <span
      class="hbox"
      style="margin-left: 0.7770em; bottom: 0.0228em;"
    >
      <span
        class="box math-i"
        style="height: 0.6833em; padding-right: 0.0715em; line-height: 0.8677em;"
      >
        K
      </span>
      <span
        class="box main-r"
        style="margin-left: 0.2222em; height: 0.6667em; line-height: 0.5357em;"
      >
        +
      </span>
      <span
        class="box math-i"
        style="margin-left: 0.2222em; height: 0.4306em; line-height: 0.3621em;"
      >
        a
      </span>
    </span>
    <span
      style="min-width: 0.853em; height: 1.0000em; background: #31121355; display: inline-block; bottom: -0.7667em;"
    >
      <svg
        height="1.0000em"
        preserveAspectRatio="xMinYMin slice"
        viewBox="75 73 400000 1000"
        width="3.4486544444444447em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M95,702
  c-2.7,0,-7.17,-2.7,-13.5,-8c-5.8,-5.3,-9.5,-10,-9.5,-14
  c0,-2,0.3,-3.3,1,-4c1.3,-2.7,23.83,-20.7,67.5,-54
  c44.2,-33.3,65.8,-50.3,66.5,-51c1.3,-1.3,3,-2,5,-2c4.7,0,8.7,3.3,12,10
  s173,378,173,378c0.7,0,35.3,-71,104,-213c68.7,-142,137.5,-285,206.5,-429
  c69,-144,104.5,-217.7,106.5,-221
  l0 -0
  c5.3,-9.3,12,-14,20,-14
  H400000v40H845.2724
  s-225.272,467,-225.272,467s-235,486,-235,486c-2.7,4.7,-9,7,-19,7
  c-6,0,-10,-1,-12,-3s-194,-422,-194,-422s-65,47,-65,47z
  M834 80h400000v40h-400000z"
        />
      </svg>
    </span>
  </span>
  <span
    class="vlist"
    style="height: 3.0000em; vertical-align: -0.2950em;"
  >
    <span
      class="hbox"
      style="margin-left: 0.9500em; bottom: 0.2950em;"
    >
      <span
        class="box size2"
        style="height: 2.2222em; padding-right: 0.4445em; line-height: 2.2220em;"
      >
        ∫
      </span>
    </span>
    <span
      style="min-width: 1.02em; height: 3.0000em; background: #31121355; display: inline-block; bottom: -2.2222em;"
    >
      <svg
        height="3.0000em"
        preserveAspectRatio="xMinYMin slice"
        viewBox="75 73 400000 3000"
        width="1.95001em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M473,2793
c339.3,-1799.3,509.3,-2700,510,-2702 l0 -0
c3.3,-7.3,9.3,-11,18,-11 H400000v40H1017.7
s-90.5,478,-276.2,1466c-185.7,988,-279.5,1483,-281.5,1485c-2,6,-10,9,-24,9
c-8,0,-12,-0.7,-12,-2c0,-1.3,-5.3,-32,-16,-92c-50.7,-293.3,-119.7,-693.3,-207,-1200
c0,-1.3,-5.3,8.7,-16,30c-10.7,21.3,-21.3,42.7,-32,64s-16,33,-16,33s-26,-26,-26,-26
s76,-153,76,-153s77,-151,77,-151c0.7,0.7,35.7,202,105,604c67.3,400.7,102,602.7,104,
606zM1001 80h400000v40H1017.7z"
        />
      </svg>
    </span>
  </span>
</span>
`;

exports[`supsub 1`] = `
<span
  class="hbox"
>
  <span
    class="hbox"
  >
    <span
      class="box math-i"
      style="height: 0.4306em; line-height: 0.3621em;"
    >
      a
    </span>
    <span
      class="vlist"
      style="height: 0.5978em; vertical-align: 0.4130em;"
    >
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em; bottom: -0.0000em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
        <span
          class="box math-i"
          style="height: 0.8540em; padding-right: 0.0572em; line-height: 0.8200em;"
        >
          j
        </span>
      </span>
    </span>
  </span>
  <span
    class="hbox"
  >
    <span
      class="box math-i"
      style="height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
    >
      f
    </span>
    <span
      class="vlist"
      style="height: 0.4783em; vertical-align: 0.4130em;"
    >
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.6833em; padding-right: 0.0715em; line-height: 0.8677em;"
        >
          K
        </span>
      </span>
    </span>
  </span>
  <span
    class="hbox"
  >
    <span
      class="box math-i"
      style="height: 0.6833em; padding-right: 0.0715em; line-height: 0.8677em;"
    >
      K
    </span>
    <span
      class="vlist"
      style="height: 0.3014em; vertical-align: -0.1500em;"
    >
      <span
        class="hbox"
        style="margin-left: -0.1022em; margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
    </span>
  </span>
  <span
    class="hbox"
  >
    <span
      class="box math-i"
      style="height: 0.4306em; line-height: 0.3621em;"
    >
      a
    </span>
    <span
      class="vlist"
      style="height: 0.6222em; vertical-align: -0.1500em;"
    >
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
        >
          f
        </span>
      </span>
    </span>
  </span>
  <span
    class="hbox"
  >
    <span
      class="box math-i"
      style="height: 0.4306em; line-height: 0.3621em;"
    >
      a
    </span>
    <span
      class="vlist"
      style="height: 1.0975em; align-items: start; vertical-align: -0.2470em;"
    >
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em; bottom: 0.0000em;"
      >
        <span
          class="box math-i"
          style="height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
        >
          f
        </span>
      </span>
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em; bottom: 0.2484em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
    </span>
  </span>
  <span
    class="hbox"
  >
    <span
      class="box math-i"
      style="height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
    >
      f
    </span>
    <span
      class="vlist"
      style="height: 0.9614em; align-items: start; vertical-align: -0.2470em;"
    >
      <span
        class="hbox"
        style="margin-left: -0.1538em; margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em; bottom: 0.5123em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
    </span>
  </span>
</span>
`;

exports[`supsub 2`] = `
<span
  class="hbox"
>
  <span
    class="hbox"
  >
    <span
      class="hbox"
    >
      <span
        class="box main-r"
        style="height: 1.0000em; line-height: 0.8690em;"
      >
        (
      </span>
      <span
        class="hbox"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
      <span
        class="box main-r"
        style="height: 1.0000em; line-height: 0.8690em;"
      >
        )
      </span>
    </span>
    <span
      class="vlist"
      style="height: 0.3014em; vertical-align: 0.5029em;"
    >
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
    </span>
  </span>
  <span
    class="hbox"
    style="margin-left: 0.1667em;"
  >
    <span
      class="hbox"
    >
      <span
        class="box main-r"
        style="height: 1.0000em; line-height: 0.8690em;"
      >
        (
      </span>
      <span
        class="hbox"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
      <span
        class="box main-r"
        style="height: 1.0000em; line-height: 0.8690em;"
      >
        )
      </span>
    </span>
    <span
      class="vlist"
      style="height: 0.3014em; vertical-align: -0.2997em;"
    >
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
    </span>
  </span>
  <span
    class="hbox"
    style="margin-left: 0.1667em;"
  >
    <span
      class="hbox"
    >
      <span
        class="box main-r"
        style="height: 1.0000em; line-height: 0.8690em;"
      >
        (
      </span>
      <span
        class="hbox"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
      <span
        class="box main-r"
        style="height: 1.0000em; line-height: 0.8690em;"
      >
        )
      </span>
    </span>
    <span
      class="vlist"
      style="height: 1.1040em; align-items: start; vertical-align: -0.2997em;"
    >
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em; bottom: 0.7160em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
    </span>
  </span>
  <span
    class="vlist"
    style="margin-left: 0.1667em; vertical-align: -1.2351em;"
  >
    <span
      class="hbox"
      style="margin-bottom: 0.1429em; margin-top: 0.4266em; font-size: 0.7000em;"
    >
      <span
        class="box math-i"
        style="height: 0.4306em; line-height: 0.3621em;"
      >
        a
      </span>
    </span>
    <span
      class="box size2"
      style="height: 1.6000em; line-height: 1.6020em;"
    >
      ∑
    </span>
    <span
      class="hbox"
      style="margin-bottom: 0.2857em; margin-top: 0.1429em; font-size: 0.7000em;"
    >
      <span
        class="box math-i"
        style="height: 0.4306em; line-height: 0.3621em;"
      >
        a
      </span>
    </span>
  </span>
  <span
    class="hbox"
  >
    <span
      class="box size2"
      style="height: 2.2222em; padding-right: 0.4445em; line-height: 2.2220em;"
    >
      ∫
    </span>
    <span
      class="vlist"
      style="height: 2.3262em; align-items: start; vertical-align: -0.9119em;"
    >
      <span
        class="hbox"
        style="margin-left: -0.6349em; margin-right: 0.0714em; font-size: 0.7000em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
      <span
        class="hbox"
        style="margin-right: 0.0714em; font-size: 0.7000em; bottom: 2.4621em;"
      >
        <span
          class="box math-i"
          style="height: 0.4306em; line-height: 0.3621em;"
        >
          a
        </span>
      </span>
    </span>
  </span>
</span>
`;

exports[`symbol box 1`] = `
<span
  class="box math-i"
  style="height: 0.8540em; padding-right: 0.0572em; line-height: 0.8200em;"
>
  j
</span>
`;

exports[`symbol box 2`] = `
<span
  class="hbox"
>
  <span
    class="box math-i"
    style="height: 0.4306em; line-height: 0.3621em;"
  >
    a
  </span>
  <span
    class="box main-r"
    style="margin-left: 0.2222em; height: 0.6667em; line-height: 0.5357em;"
  >
    +
  </span>
  <span
    class="box math-i"
    style="margin-left: 0.2222em; height: 0.8889em; padding-right: 0.1076em; line-height: 0.8899em;"
  >
    f
  </span>
  <span
    class="box main-r"
    style="margin-left: 0.2778em; height: 0.2337em; line-height: 0.1027em;"
  >
    =
  </span>
  <span
    class="box size2"
    style="margin-left: 0.2778em; height: 2.2222em; padding-right: 0.4445em; line-height: 2.2220em;"
  >
    ∫
  </span>
</span>
`;
